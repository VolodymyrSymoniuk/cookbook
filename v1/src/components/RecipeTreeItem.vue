<template>
  <li>
    <div
      :class="[isFolder ? 'bold' : '', 'item', 'pointer']"
      @click="toggle"
      @dblclick="makeFolder"
    >
      {{ item.description }}
      <span v-if="isFolder">[{{ isOpen ? "-" : "+" }}]</span>
      <span
        v-if="!isRoot"
        @click.stop="displayRecipe"
        class="material-icons show-icon"
        :class="{ active: isRecipeActive }"
      >
        visibility
      </span>
    </div>
    <ul v-show="isOpen" v-if="isFolder" class="children">
      <RecipeTreeItem
        v-for="(childId, index) in item.children"
        :key="index"
        :recipeId="childId"
      ></RecipeTreeItem>
      <li class="pointer add-item" @click="addItem">+</li>
    </ul>
  </li>
</template>

<script>
import shortid from "shortid";
import { mapMutations, mapActions } from "vuex";
import RecipeTreeItem from "@/components/RecipeTreeItem";

export default {
  name: "RecipeTreeItem",
  components: { RecipeTreeItem },
  props: {
    recipeId: String
  },
  data: () => ({
    isOpen: false
  }),
  mounted() {
    if (this.isRoot) {
      this.isOpen = true;
    }
  },
  computed: {
    item() {
      return this.$store.getters.getRecipeById(this.recipeId);
    },
    isFolder() {
      return this.item.children && this.item.children.length;
    },
    isRoot() {
      return this.recipeId === this.$store.getters.rootRecipeId;
    },
    isRecipeActive() {
      return this.$store.getters.isRecipeActive(this.recipeId);
    }
  },
  methods: {
    ...mapMutations(["setActiveRecipeById", "makeRecipeParent"]),
    ...mapActions(["addNewRecipe"]),
    toggle() {
      if (this.isFolder) {
        this.isOpen = !this.isOpen;
      }
    },
    makeFolder() {
      if (!this.isFolder) {
        this.makeRecipeParent(this.item);
        this.addItem(this.item);
        this.isOpen = true;
      }
    },
    async addItem() {
      let { id: parent, description } = this.item;
      if (parent === "root") {
        try {
          description = await this.$prompt("Enter description of new recipe:");
        } catch (_) {
          this.$alert("Invalid recipe description entered!");
          return;
        }
      }

      this.addNewRecipe({
        id: shortid.generate(),
        description,
        versions: [],
        parent,
        createdAt: Date.now()
      });
    },
    displayRecipe() {
      this.setActiveRecipeById(this.recipeId);
    }
  }
};
</script>
