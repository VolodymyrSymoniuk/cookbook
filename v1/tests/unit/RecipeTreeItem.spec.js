import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import RecipeTreeItem from "@/components/RecipeTreeItem";

const localVue = createLocalVue();
localVue.use(Vuex);

const currentRecipe = {
  id: "2WEKaVNO",
  parent: "root",
  description: "Fried potato",
  versions: ["Potato", "Pomamo", "Posestro"],
  createdAt: 1610702654321,
  children: []
};

let activeRecipeId = "not current recipe id";

describe("RecipeTree component", () => {
  let mutations, actions, getters, store;

  beforeEach(() => {
    // clear current recipe children
    currentRecipe.children = [];

    getters = {
      rootRecipeId: () => "root",
      getRecipeById: () => () => currentRecipe,
      isRecipeActive: () => () => currentRecipe.id === activeRecipeId
    };
    actions = {
      addNewRecipe: jest.fn()
    };
    mutations = {
      setActiveRecipeById: jest.fn(() => {
        activeRecipeId = currentRecipe.id;
      }),
      makeRecipeParent: jest.fn(() => {
        currentRecipe.children.push("childId");
      })
    };
    store = new Vuex.Store({
      actions,
      getters,
      mutations
    });
  });

  it("should render as expected", () => {
    const wp = shallowMount(RecipeTreeItem, { store, localVue });
    expect(wp).toMatchSnapshot();
  });

  it("should make this TreeItem active", async () => {
    const wp = shallowMount(RecipeTreeItem, { store, localVue });
    wp.find(".show-icon").trigger("click");
    expect(mutations.setActiveRecipeById).toHaveBeenCalled();
  });

  it("should add recipe on '+' item click", async () => {
    store.commit("makeRecipeParent");
    const wp = shallowMount(RecipeTreeItem, {
      store,
      localVue,
      propsData: () => ({ isOpen: true })
    });
    wp.find(".add-item").trigger("click");
    expect(actions.addNewRecipe).toHaveBeenCalled();
  });

  it("correctly toggles on click", async () => {
    store.commit("makeRecipeParent");
    const methods = { toggle: jest.fn() };
    const wp = shallowMount(RecipeTreeItem, { store, localVue, methods });

    expect(wp.find(".children").isVisible()).toBeFalsy();
    wp.find(".item").trigger("click");
    expect(methods.toggle).toHaveBeenCalled();
  });

  it("should create children branch on double click", () => {
    const wp = shallowMount(RecipeTreeItem, { store, localVue });
    wp.find(".item").trigger("dblclick");
    expect(mutations.makeRecipeParent).toHaveBeenCalled();
    expect(actions.addNewRecipe).toHaveBeenCalled();
  });
});
