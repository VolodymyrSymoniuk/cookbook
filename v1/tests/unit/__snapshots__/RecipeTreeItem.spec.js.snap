// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeTree component should render as expected 1`] = `
<li>
  <div class="item pointer">
    Fried potato
    <!----> <span class="material-icons show-icon">
      visibility
    </span>
  </div>
  <!---->
</li>
`;
